### PR TITLE
[feat][authentication] Add JWKS support for AuthenticationProviderToken

### PR DESCRIPTION
**Pull request by nodece**
Fix https://github.com/apache/pulsar/issues/8152

### Motivation

Add JWKS support for AuthenticationProviderToken.

### Modifications

- Add `tokenKeySet` config to provide the JWKS support
 
### Verifying this change

- Added RSA test
- Added EC test

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/nodece/pulsar/pull/11

### CODE REVIEW

**Review by Demogorgon314 — APPROVED** [review]

**Review by rdhabalia — CHANGES_REQUESTED** [review]
this PR is adding new feature and it should go through PIP.

**Comment by codelipenghui** [comment]
> this PR is adding new feature and it should go through PIP.

+1

@nodece  Could you please start a proposal for this change?

**Comment by nodece** [comment]
> this PR is adding new feature and it should go through PIP.

Thanks.

Let me make a PIP.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java [hunk 1]
```diff
@@ -409,4 +444,134 @@ public String getHeader(String name) {
...
+        private final Map<String, Key> keyMap = new HashMap<>();
+
+        public Jwk(String data) {
+            String json;
+            try {
+                byte[] bytes = AuthTokenUtils.readKeyFromUrl(data);
+                if (bytes == null || bytes.length == 0) {
+                    throw new IOException("invalid JWKs data");
+                }
+                json = new String(AuthTokenUtils.readKeyFromUrl(data), StandardCharsets.UTF_8);
```
Comment (tuteng) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:491]: Why call this function `AuthTokenUtils.readKeyFromUrl(data)` again?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java [hunk 2]
```diff
@@ -266,7 +301,7 @@ private Key getValidationKey(ServiceConfiguration conf) throws IOException {
             final byte[] validationKey = AuthTokenUtils.readKeyFromUrl(tokenPublicKey);
             return AuthTokenUtils.decodePublicKey(validationKey, publicKeyAlg);
         } else {
-            throw new IOException("No secret key was provided for token authentication");
+            return null;
```
Comment (tuteng) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:304]: I don't think it's the third case, it seems like it should be considered adding it to the if branch of the publickey, which is essentially where the public key is read (file, string or an http host)
I think the `TokenSigningKeyResolver` should be called in the `readKeyFromUrl` function

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java [hunk 3]
```diff
@@ -409,4 +444,134 @@ public String getHeader(String name) {
             return super.getHeader(name);
         }
     }
+
+    private static final class TokenSigningKeyResolver implements SigningKeyResolver {
```
Comment (tuteng) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:448]: I think the `TokenSigningKeyResolver` should be called in the `readKeyFromUrl` function

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java (hunk 2 above)
Comment (nodece) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:304]: See lines 158-162.

This method just gets the public or secret key. When returns null, we read the JWKS data.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java (hunk 1 above)
Comment (nodece) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:491]: Good catch. Here should be `bytes`.

**Comment by nodece** [comment]
Ping @michaeljmarshall, could you track this PR?

**Review by michaeljmarshall — COMMENTED** [review]
https://github.com/apache/pulsar/issues/8152 asks for support to rotate public keys and to load new public keys "on the fly". This PR appears to add support for loading multiple public keys on start up, but it does not provide support for rotation or dynamic loading. Is that correct? If so, I don't think we should say this PR fixes https://github.com/apache/pulsar/issues/8152.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java [hunk 4]
```diff
@@ -409,4 +445,134 @@ public String getHeader(String name) {
...
+            }
+
+            Iterator<JsonNode> elements = keysNode.elements();
+            while (elements.hasNext()) {
+                JsonNode node = elements.next();
+                String type = node.get("kty").textValue();
+                String kid = node.get("kid").textValue();
+                KeyFactory kf;
+                // Reference from:
+                // https://github.com/auth0/jwks-rsa-java/blob/0.21.2/src/main/java/com/auth0/jwk/Jwk.java#L176
```
[the remark this answers is not in the document]
Comment (nodece) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:524]: We just use the `Jwk` class, and don't want to introduce the other dependency.

**Comment by nodece** [comment]
> #8152 asks for support to rotate public keys and to load new public keys "on the fly". This PR appears to add support for loading multiple public keys on start up, but it does not provide support for rotation or dynamic loading. Is that correct? If so, I don't think we should say this PR fixes #8152.

You are right, I think we can listener to the file changes.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaeljmarshall) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java:524]: The license might require that we change this PR: https://github.com/auth0/jwks-rsa-java/blob/0.21.2/LICENSE.

@dave2wave - are you able to provide any guidance on the licensing aspect here? Thanks!

**Review by michaeljmarshall — CHANGES_REQUESTED** [review]
Requesting changes while we determine the licensing portion of this PR.

**Comment by nodece** [comment]
Use https://github.com/apache/pulsar/pull/22215 instead of this PR.

**Pull request closed without merging**